numpy/core/_internal.py

```diff
@@ -287,23 +287,45 @@ def _newnames(datatype, order):
         return tuple(list(order) + nameslist)
     raise ValueError("unsupported order value: %s" % (order,))
 
+def isstring(obj):
+    try:
+        #py2
+        return isinstance(obj, basestring)
+    except NameError:
+        #py3
+        return isinstance(obj, str)
+
 # Given an array with fields and a sequence of field names
-# construct a new array with just those fields copied over
-def _index_fields(ary, fields):
-    from .multiarray import empty, dtype, array
+# construct new array with just those fields
+def _index_fields(ary, names):
+    from .multiarray import dtype, array
     dt = ary.dtype
 
-    names = [name for name in fields if name in dt.names]
-    formats = [dt.fields[name][0] for name in fields if name in dt.names]
-    offsets = [dt.fields[name][1] for name in fields if name in dt.names]
+    if ary.dtype.fields == None:
+        raise ValueError("array has no fields")
+
+    #use getfield to index a single field
+    if isstring(names):
+        try:
+            return ary.getfield(dt.fields[names][0], dt.fields[names][1])
+        except KeyError as e:
+            # outside code expects valueerror in field lookup failure
+            raise ValueError(e)
 
-    view_dtype = {'names':names, 'formats':formats, 'offsets':offsets, 'itemsize':dt.itemsize}
-    view = ary.view(dtype=view_dtype)
+    for name in names:
+        if name not in dt.fields:
+            raise ValueError("field name %s not found" % name)
```

Comment:
Oh, right it would be KeyError. I think we can probably change it to be honest. I changed a couple of weird value to index error last version and nobody seemed to have noticed.